File: graphs/java/Graphs/src/Graph.java
/**
 * Created by petergoldsborough on 09/30/15.
 */

import java.util.ArrayList;

/*

- `Graph(int V)`: create an empty graph with `V` vertices.
- `Graph(File in)`: create graph from an input string.
- `void addEdge(int v, int w)`: add an edge between `v` and `w`.
- `Iterable<Integer> adjacent(int v)`: Get an iterator to all vertices adjacent to `v`.
- `int numberOfEdges()`: The number of edges in the graph.
- `int numberOfVertices()`: the number of vertices in the graph.

 */

public class Graph
{
	public static class Edge
	{
		public Edge(int vertex, int id)
		{
			this.vertex = vertex;
			this.id = id;
		}
		
		public int hashCode()
		{
			return vertex ^ id;
		}
		
		public boolean equals(Object object)
		{
			if (! (object instanceof Edge)) return false;
			
			Edge other = (Edge) object;

			return this.id == other.id;
		}
		
		public final int vertex;
		public final int id;
	}
	
	public Graph(int numberOfVertices)
	{
		this.vertices = (ArrayList<Edge>[]) new ArrayList[numberOfVertices];

		for (int vertex = 0; vertex < numberOfVertices; ++vertex)
		{
			vertices[vertex] = new ArrayList<>();
		}
	}

	public void addEdge(int first, int second)
	{
		assertIndex(first);
		assertIndex(second);

		vertices[first].add(new Edge(second, edges));
		vertices[second].add(new Edge(first, edges));

		++edges;
	}

	ArrayList<Edge> adjacent(int vertex)
	{
		assertIndex(vertex);

		return vertices[vertex];
	}



	public int numberOfEdges()
	{
		return edges;
	}

	public int numberOfVertices()
	{
		return vertices.length;
	}

	private void assertIndex(int vertex)
	{
		if (vertex < 0 || vertex > vertices.length)
		{
			throw new IllegalArgumentException("Vertex index out of range!");
		}
	}

	private int edges = 0;

	private ArrayList<Edge>[] vertices;

}

File: graphs/cpp/Graphs/src/Graph.cpp
//
//  Graph.cpp
//  Graphs
//
//  Created by Peter Goldsborough on 09/30/15.
//  Copyright © 2015 Peter Goldsborough. All rights reserved.
//

#include <queue>

#include "Graph.hpp"

Graph::Graph(vertex_t vertices)
: _vertices(vertices)
, _edges(0)
{ }

void Graph::add_edge(vertex_t first, vertex_t second)
{
	_vertices[first].push_back({second, _edges});
	_vertices[second].push_back({first, _edges});
	
	++_edges;
}

const Graph::adjacent_t& Graph::adjacent(vertex_t vertex) const
{
	return _vertices[vertex];
}

Graph::size_t Graph::number_of_edges() const
{
	return _edges;
}

Graph::size_t Graph::number_of_vertices() const
{
	return _vertices.size();
}